File: arai60/find-k-pairs-with-smallest-sums/step2.cpp
class Solution {
public:
    vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2, int k) {
        priority_queue<SumPair> min_sum_pairs;
        int n = nums1.size();
        int m = nums2.size();
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                int sum = nums1[i] + nums2[j];
                if (min_sum_pairs.size() < k) {
                    min_sum_pairs.push({sum, {nums1[i], nums2[j]}});
                } else if (min_sum_pairs.size() == k && sum < min_sum_pairs.top().sum) {
                    min_sum_pairs.pop();
                    min_sum_pairs.push({sum, {nums1[i], nums2[j]}});
                } else {
                    break;
                }
            }
        }

        vector<vector<int>> answer;
        while (!min_sum_pairs.empty()) {
            answer.push_back(min_sum_pairs.top().pair);
            min_sum_pairs.pop();
        }
        return answer;
    }

private:
    struct SumPair {
        int sum;
        vector<int> pair;

        bool operator<(const SumPair& other) const {
            return sum < other.sum;
        }
    };
};

File: arai60/intersection-of-two-arrays/step2.py
class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1.sort()
        nums2.sort()

        index_nums1 = 0
        index_nums2 = 0
        answer = []
        while index_nums1 < len(nums1) and index_nums2 < len(nums2):
            if nums1[index_nums1] < nums2[index_nums2]:
                index_nums1 += 1
                continue
            if nums1[index_nums1] > nums2[index_nums2]:
                index_nums2 += 1
                continue

            common = nums1[index_nums1]
            answer.append(common)
            while index_nums1 < len(nums1) and nums1[index_nums1] == common:
                index_nums1 += 1
            while index_nums2 < len(nums2) and nums2[index_nums2] == common:
                index_nums2 += 1

        return answer
